# Курсовая работа № 2
## Аналитика. Начальный уровень

Параметры для подключения следующие: хост – `dsstudents.skillbox.ru`, порт – `5432`, имя базы данных – `human_resources`, пользователь – `readonly`, пароль – `6hajV34RTQfmxhS`. Таблицы, доступные для анализа, – `hr_dataset`, `production_staff`, `recruiting_costs`, `salary_grid`.

In [1]:
import pandas as pd
import sqlalchemy
import psycopg2

In [2]:
conn = 'postgresql+psycopg2://readonly:6hajV34RTQfmxhS@dsstudents.skillbox.ru:5432/human_resources'

engine = sqlalchemy.create_engine(conn)

connect = engine.connect()

In [3]:
inspector = sqlalchemy.inspect(engine)
inspector.get_table_names()

['hr_dataset', 'production_staff', 'recruiting_costs', 'salary_grid']

In [4]:
hr_dataset = pd.read_sql('select * from hr_dataset', connect)
hr_dataset.head()

id         Employee Name  Employee Number  marriedid  maritalstatusid  \
0   1            Brown, Mia       1103024456          1                1   
1   2  LaRotonda, William         1106026572          0                2   
2   3      Steans, Tyrone         1302053333          0                0   
3   4       Howard, Estelle       1211050782          1                1   
4   5           Singh, Nan        1307059817          0                0   

   genderid  empstatus_id  deptid  perf_scoreid  age  ...  Date of Hire  \
0         0             1       1             3   30  ...    2008-10-27   
1         1             1       1             3   34  ...    2014-01-06   
2         1             1       1             3   31  ...    2014-09-29   
3         0             1       1             9   32  ...    2015-02-16   
4         0             1       1             9   30  ...    2015-05-01   

  Days Employed  Date of Termination       Reason For Term Employment Status  \
0          3317                 None  N/A - still employed            Active   
1          1420                 None  N/A - still employed            Active   
2          1154                 None  N/A - still employed            Active   
3            58           2015-04-15  N/A - still employed            Active   
4           940                 None  N/A - still employed            Active   

      department                  position        Manager Name  \
0  Admin Offices              Accountant I  Brandon R. LeBlanc   
1  Admin Offices              Accountant I  Brandon R. LeBlanc   
2  Admin Offices              Accountant I  Brandon R. LeBlanc   
3  Admin Offices  Administrative Assistant  Brandon R. LeBlanc   
4  Admin Offices  Administrative Assistant  Brandon R. LeBlanc   

          Employee Source         Performance Score  
0      Diversity Job Fair               Fully Meets  
1      Website Banner Ads               Fully Meets  
2         Internet Search               Fully Meets  
3  Pay Per Click - Google  N/A- too early to review  
4      Website Banner Ads  N/A- too early to review  

[5 rows x 29 columns]

In [5]:
production_staff = pd.read_sql('select * from production_staff', connect)
production_staff.head()

id       Employee Name Race Desc Date of Hire    TermDate  \
0   1   Albert, Michael       White   2011-08-01        None   
1   2      Bozzi, Charles     Asian   2013-09-30  2014-08-07   
2   3  Butler, Webster  L     White   2016-01-28        None   
3   4         Dunn, Amy       White   2014-09-18        None   
4   5     Gray, Elijiah       White   2015-06-02        None   

        Reason for Term       Employment Status         Department  \
0  N/A - still employed                  Active  Production          
1              retiring  Voluntarily Terminated  Production          
2  N/A - still employed                  Active  Production          
3  N/A - still employed                  Active  Production          
4  N/A - still employed                  Active  Production          

             Position     Pay    Manager Name Performance Score  \
0  Production Manager  $54.50  Elisa Bramante       Fully Meets   
1  Production Manager  $50.50  Elisa Bramante       Fully Meets   
2  Production Manager  $55.00  Elisa Bramante           Exceeds   
3  Production Manager  $51.00  Elisa Bramante       Fully Meets   
4  Production Manager  $54.00  Elisa Bramante       Fully Meets   

   Abutments/Hour Wk 1  Abutments/Hour Wk 2  Daily Error Rate  \
0                  0.0                  0.0               0.0   
1                  0.0                  0.0               0.0   
2                  0.0                  0.0               0.0   
3                  0.0                  0.0               0.0   
4                  0.0                  0.0               0.0   

   90-day Complaints  
0                0.0  
1                0.0  
2                0.0  
3                0.0  
4                0.0

In [6]:
recruiting_costs = pd.read_sql('select * from recruiting_costs', connect)
recruiting_costs.head()

id           Employment Source  January   February  March  April  May  \
0   1                   Billboard       520       520    520    520    0   
1   2               Careerbuilder       410       410    410    820  820   
2   3  Company Intranet - Partner         0         0      0      0    0   
3   4          Diversity Job Fair         0      5129      0      0    0   
4   5           Employee Referral         0         0      0      0    0   

   June  July  August  September  October  November  December  Total  
0     0   612     612        729      749       910       500   6192  
1   410   410     820        820     1230       820       410   7790  
2     0     0       0          0        0         0         0      0  
3     0     0    4892          0        0         0         0  10021  
4     0     0       0          0        0         0         0      0

In [7]:
salary_grid = pd.read_sql('select * from salary_grid', connect)
salary_grid.head(20)

id                      Position  Salary Min  Salary Mid  Salary Max  \
0    1      Administrative Assistant       30000       40000       50000   
1    2  Sr. Administrative Assistant       35000       45000       55000   
2    3                  Accountant I       42274       51425       62299   
3    4                 Accountant II       50490       62158       74658   
4    5                Sr. Accountant       63264       76988       92454   
5    6              Network Engineer       50845       66850       88279   
6    7          Sr. Network Engineer       79428       99458      120451   
7    8        Database Administrator       50569       68306       93312   
8    9                       Sr. DBA       92863      116007      139170   
9   10       Production Technician I       30000       40000       50000   
10  11      Production Technician II       38000       48000       58000   
11  12    Lead Production Technician       45000       55000       65000   

    Hourly Min  Hourly Mid  Hourly Max  
0        14.42       19.23       24.04  
1        16.83       21.63       26.44  
2        20.32       24.72       29.95  
3        24.27       29.88       35.89  
4        30.42       37.01       44.45  
5        24.44       32.14       42.44  
6        38.19       47.82       57.91  
7        24.31       32.84       44.86  
8        44.65       55.77       66.91  
9        14.42       19.23       24.04  
10       18.27       23.08       27.88  
11       21.63       26.44       31.25

### Для начала проведем небольшой анализ имеющихся данных

In [8]:
# Посчитаем общее количество сотрудников
pd.read_sql('SELECT COUNT(hr."Employee Name") FROM hr_dataset AS hr', connect)

count
0    310

In [9]:
# Проверим статус сотрудников
pd.read_sql('SELECT hr."Employment Status" AS status, COUNT (hr."Employee Name")\
             FROM hr_dataset AS hr GROUP BY status', connect)

status  count
0            Future Start     11
1  Voluntarily Terminated     88
2                  Active    183
3    Terminated for Cause     14
4        Leave of Absence     14

In [10]:
# Посчитаем количество сотрудников в отделах
pd.read_sql('SELECT hr."department" AS dept, COUNT(hr."Employee Name")\
             FROM hr_dataset AS hr GROUP BY dept', connect)

dept  count
0     Production           208
1         Admin Offices     10
2                 IT/IS     50
3      Executive Office      1
4  Software Engineering     10
5                 Sales     31

In [11]:
# Посчитаем количество действующих сотрудников в отделах
pd.read_sql('SELECT hr."department" AS dept, COUNT(hr."Employee Name")\
             FROM hr_dataset AS hr\
             WHERE (hr."Employment Status" =\'Active\' OR "Employment Status" = \'Leave of Absence\') GROUP BY dept', connect)

dept  count
0     Production           117
1         Admin Offices      8
2                 IT/IS     38
3      Executive Office      1
4  Software Engineering      7
5                 Sales     26

In [12]:
# Проверим гендерный состав сотрудников
pd.read_sql('SELECT hr."sex" AS sex,\
             COUNT (hr."Employee Name") FROM hr_dataset AS hr GROUP BY sex', connect)

sex  count
0  Female    177
1    Male    133

In [13]:
# Выясним расовый состав сотрудников
pd.read_sql('SELECT hr."racedesc" AS race, COUNT(hr."racedesc") AS count\
             FROM hr_dataset AS hr\
             WHERE (hr."Employment Status" =\'Active\' OR "Employment Status" = \'Leave of Absence\')\
             GROUP BY race ORDER BY count DESC', connect)

race  count
0                             White    120
1         Black or African American     38
2                             Asian     22
3                 Two or more races     10
4  American Indian or Alaska Native      4
5                          Hispanic      3

In [14]:
# Выясним минимальный,максимальный и средний возраст сотрудников
pd.read_sql('SELECT MIN(hr."age") AS min_age, ROUND(AVG(hr."age")) AS avg_age, MAX(hr."age") AS max_age  \
             FROM hr_dataset AS hr\
             WHERE (hr."Employment Status" = \'Active\' OR hr."Employment Status" = \'Leave of Absence\')', connect)

min_age  avg_age  max_age
0       25     38.0       67

In [15]:
# Проверим соответствие занимаемой должности
pd.read_sql('SELECT hr."Performance Score" AS perf_score, COUNT(hr."Employee Name") AS count\
             FROM hr_dataset AS hr\
             WHERE (hr."Employment Status" = \'Active\' OR hr."Employment Status" = \'Leave of Absence\')\
             GROUP BY perf_score ORDER BY count DESC', connect)

perf_score  count
0               Fully Meets    125
1              90-day meets     18
2                   Exceeds     18
3  N/A- too early to review     15
4               Exceptional      9
5         Needs Improvement      7
6                       PIP      5

In [16]:
# Узнаем количество ежедневных ошибок 
pd.read_sql('SELECT SUM(ps."Daily Error Rate") AS error FROM production_staff AS ps', connect)

error
0    188

In [17]:
# Количество жалоб за 90 дней
pd.read_sql('SELECT SUM(ps."90-day Complaints") AS complaints FROM production_staff AS ps', connect)

complaints
0          66

In [18]:
# Выясним минимальную, максимальную и среднюю продолжительность работы сотрудников
pd.read_sql('SELECT MIN(hr."Days Employed") AS min_days, MAX(hr."Days Employed") AS max_days,\
             ROUND(AVG(hr."Days Employed")) AS avg_days\
             FROM hr_dataset AS hr', connect)

min_days  max_days  avg_days
0         2      4339    1296.0

Проведя предварительный анализ, я выявил неточности и несоответствия в предоставленных данных. Часть сотрудников в разных        таблицах числятся в разных отделах:<br>
   - 'King, Janet' в таблице 'production_staff' имеет статус 'Active' в отделе 'Production' и занимает должность 'Production Technician I', а в таблице 'hr_dataset' она имеет статус 'Active' в отделе 'Executive Office' и занимает должность 'President & CEO'. 
   -  'Sweetwater, Alex' имеет статус 'Active' в отделе 'Production' и занимает должность 'Production Technician II', а в таблице 'hr_dataset' он имеет статус 'Active' в отделе 'Software Engineering' и занимает должность 'Software Engineering Manager'. 
   - 'Bramante, Elisa' по данным таблицы 'hr_dataset' числится в отделе 'Production', но отсутствует в таблице               'production_staff'.<br>

Также есть несоответсвие в таблцие 'salary_grid'. Отсутствуют либо имеют другое название такие должности как:
   - 'Production Manager' в отделе 'Production'
   - 'IT Support' в отделе 'IT/IS' и др.<br>

Стоит обратить внимание на важность корректного составления и заполнения этих таблиц.<br>

Были предоставлены данные о 310 сотрудниках компании. Из них:<br>
   - 197 человек, это действующие сотрудники
   - 11 человек приступят к работе в будущем
   - 102 человека уже покинули компанию.<br>
   
Большая часть сотрудников работает в отделе 'Production' - 117 человек.    
Среди сотрудников преобладают люди белой расы - 120 человек.<br>
 
В полученных результатах мое внимание привлекли:
   - 'Daily Error Rate' - Количество ошибок в день(188 ошибок)
   - 'Voluntarily Terminated' - Увольнение по собственному желанию (88 увольнений)
   - 'Days Employed' - Количество отработаных дней (средняя продолжительность работы - 1296 дней).<br>
   
Эти данные станут предметом исследования.

### Предмет исследования №1:
### Daily Error Rate - Количество ошибок в день

In [19]:
# Зависимость должности и количества ошибок 
pd.read_sql('SELECT ps."Position" AS position, COUNT(ps."Employee Name") AS ppl_count, SUM(ps."Daily Error Rate") AS error\
             FROM production_staff as ps\
             WHERE ps."Daily Error Rate" > 0\
             GROUP BY position ORDER BY error DESC', connect)

position  ppl_count  error
0   Production Technician I         82    138
1  Production Technician II         30     50

In [20]:
# Зависимость количества ошибок от расы сотрудника
pd.read_sql('SELECT ps."Race Desc" AS race, COUNT(ps."Race Desc") AS ppl_count, SUM(ps."Daily Error Rate") AS count\
             FROM production_staff as ps\
             GROUP BY race ORDER BY count DESC', connect)

race  ppl_count  count
0                              None          0    NaN
1                             White        140  116.0
2         Black or African American         31   42.0
3                             Asian         22   15.0
4                 Two or more races         10   10.0
5                          Hispanic          3    4.0
6  American Indian or Alaska Native          3    1.0

In [44]:
# Зависимость кто у сотрудника менеджер и количество ошибок
pd.read_sql('SELECT ps."Manager Name" AS manager,\
             COUNT(ps."Employee Name") AS ppl_count,\
             SUM(ps."Daily Error Rate") AS err_count\
             FROM production_staff AS ps\
             WHERE ps."Daily Error Rate" > 0\
             GROUP BY manager ORDER BY err_count DESC', connect)

manager  ppl_count  err_count
0    Elijiah Gray         15         28
1   David Stanley         14         25
2  Webster Butler         12         24
3  Michael Albert         12         21
4   Ketsia Liebig         13         20
5        Amy Dunn         13         20
6  Kissy Sullivan         13         19
7  Brannon Miller          9         16
8   Kelley Spirea         11         15

In [43]:
# Зависимость зарплаты и кол-ва ошибок на должности 'Production Technician I'
pd.read_sql('SELECT ps."Employee Name" AS name, ps."Daily Error Rate" AS err_rate, ps."Pay" AS pay, sg."Hourly Mid" AS mid_pay\
             FROM production_staff AS ps\
             JOIN salary_grid AS sg\
             ON sg."Position" = ps."Position"\
             WHERE (sg."Position" = \'Production Technician I\' AND ps."Daily Error Rate" > 0)\
             ORDER BY err_rate DESC LIMIT 10',connect)

name  err_rate     pay  mid_pay
0          Gilles, Alex         4  $15.00    19.23
1        DiNocco, Lily          4  $22.00    19.23
2      Wallace, Theresa         4  $15.00    19.23
3        Maurice, Shana         4  $20.00    19.23
4          Jung, Judy           3  $21.00    19.23
5     Estremera, Miguel         3  $17.00    19.23
6       Desimone, Carl          3  $20.00    19.23
7      Chivukula, Enola         3  $21.00    19.23
8      Anderson, Carol          3  $16.00    19.23
9  Baczenski, Rachael           3  $17.00    19.23

In [23]:
# Зависимость зарплаты и кол-ва ошибок на должности 'Production Technician II'
pd.read_sql('SELECT ps."Employee Name" AS name, ps."Daily Error Rate" AS err_rate, ps."Pay" AS pay, sg."Hourly Mid" AS mid_pay\
             FROM production_staff AS ps\
             JOIN salary_grid AS sg\
             ON sg."Position" = ps."Position"\
             WHERE (sg."Position" = \'Production Technician II\' AND ps."Daily Error Rate" > 0)\
             ORDER BY err_rate DESC LIMIT 10',connect)

name  err_rate     pay  mid_pay
0         Oliver, Brooke          4  $25.00    23.08
1          Lunquist, Lisa         4  $26.10    23.08
2             Close, Phil         3  $26.00    23.08
3          Foreman, Tanya         3  $24.00    23.08
4           Demita, Carla         3  $29.00    23.08
5         Nowlan, Kristie         3  $26.39    23.08
6        McCarthy, Brigit         2  $26.00    23.08
7          Hunts, Julissa         2  $25.00    23.08
8        Beak, Kimberly           2  $27.00    23.08
9  Homberger, Adrienne  J         2  $28.00    23.08

In [24]:
# Зависимость пола и количества ошибок
pd.read_sql('SELECT hr."sex" AS sex, COUNT(hr."Employee Name") AS ppl_count, SUM(ps."Daily Error Rate") AS err_count\
             FROM hr_dataset AS hr\
             JOIN production_staff AS ps\
             ON ps."Employee Name" = hr."Employee Name"\
             WHERE ps."Daily Error Rate" > 0\
             GROUP BY sex',connect)

sex  ppl_count  err_count
0  Female         77        133
1    Male         35         55

### Вывод:
*Для корректности сравнения будем учитывать количество ошибок на одного человека, а не общее количество.<br>

Не выявлено никаких зависимостей количества ошибок от: 
   - должность сотрудника - около 1.6 ошибок на человека на всех должностях
   - пол сотрудника - около 1.7 ошибок на человека у обоих полов
   - зарплата сотрудника.
      
Была выявлена зависимость количества ошбиок от таких показателей как:
   - раса сотрудника: 
       - Black or African American - 1.35 ошибок на человека<br>
   - менеджер сотрудника:
      - Webster Butler - 2 ошибки на человека
      - Elijiah Gray - 1.86 ошибок на человека<br>
      
И стоит отметить *Kelley Spirea* - 1.36 ошибок на человека (минимальный показатель)

### Предмет исследования №2:
### Voluntarily Terminated - Увольнение по собственному желанию

In [26]:
# Причины увольнений
pd.read_sql('SELECT hr."Reason For Term" AS reason, COUNT(hr."Employee Name") AS count\
             FROM hr_dataset AS hr\
             WHERE hr."Employment Status" = \'Voluntarily Terminated\'\
             GROUP BY reason ORDER BY count DESC', connect)

reason  count
0                   Another position     20
1                            unhappy     14
2                         more money     11
3                      career change      9
4                              hours      8
5             relocation out of area      5
6                   return to school      5
7                           military      4
8                           retiring      4
9   maternity leave - did not return      3
10                    medical issues      3
11                       performance      1
12                        attendance      1

In [36]:
# Зависимость увольнений от отдела 
pd.read_sql('SELECT hr."department" AS dept, COUNT(hr."Employee Name") AS count\
             FROM hr_dataset AS hr\
             WHERE hr."Employment Status" = \'Voluntarily Terminated\'\
             GROUP BY dept ORDER BY count DESC', connect)

dept  count
0     Production            75
1                 IT/IS      6
2                 Sales      3
3         Admin Offices      2
4  Software Engineering      2

In [37]:
# Источнк найма и кол-во увольнений
pd.read_sql('SELECT hr."Employee Source" AS emp_sour, COUNT(hr."Employee Name") AS count\
             FROM hr_dataset AS hr\
             WHERE hr."Employment Status" = \'Voluntarily Terminated\'\
             GROUP BY emp_sour ORDER BY count DESC', connect)

emp_sour  count
0                       Diversity Job Fair     16
1        Search Engine - Google Bing Yahoo     14
2                              Monster.com     10
3   Social Networks - Facebook Twitter etc      6
4                       Newspager/Magazine      5
5                            Word of Mouth      5
6                                Billboard      4
7                                Glassdoor      4
8                                 MBTA ads      3
9                          Vendor Referral      3
10                  Pay Per Click - Google      3
11                                   Other      3
12                       Employee Referral      2
13                         Internet Search      2
14                    Professional Society      2
15              Company Intranet - Partner      1
16                           Pay Per Click      1
17                      Website Banner Ads      1
18                    On-campus Recruiting      1
19                     Information Session      1
20                 On-line Web application      1

In [29]:
# Общее количество нанятых 
pd.read_sql('SELECT hr."Employee Source" AS emp_sour, COUNT(hr."Employee Name") AS count\
             FROM hr_dataset AS hr\
             GROUP BY emp_sour ORDER BY count DESC', connect)

emp_sour  count
0                        Employee Referral     31
1                       Diversity Job Fair     29
2        Search Engine - Google Bing Yahoo     25
3                              Monster.com     24
4                   Pay Per Click - Google     21
5                     Professional Society     20
6                       Newspager/Magazine     18
7                                 MBTA ads     17
8                                Billboard     16
9                          Vendor Referral     15
10                               Glassdoor     14
11                           Word of Mouth     13
12                      Website Banner Ads     13
13                    On-campus Recruiting     12
14  Social Networks - Facebook Twitter etc     11
15                                   Other      9
16                                  Indeed      8
17                         Internet Search      6
18                     Information Session      4
19              Company Intranet - Partner      1
20                           Careerbuilder      1
21                           Pay Per Click      1
22                 On-line Web application      1

In [35]:
# Зависимость кто менеджер и кол-во увольнений
pd.read_sql('SELECT ps."Manager Name" AS manager, COUNT(ps."Employee Name") AS count\
             FROM production_staff AS ps\
             WHERE ps."Employment Status" = \'Voluntarily Terminated\'\
             GROUP BY manager ORDER BY count DESC', connect)

manager  count
0        Amy Dunn     13
1  Webster Butler     13
2  Kissy Sullivan     10
3  Michael Albert      8
4    Elijiah Gray      7
5   David Stanley      6
6  Brannon Miller      6
7   Kelley Spirea      4
8  Elisa Bramante      4
9   Ketsia Liebig      4

In [31]:
# Общее кол-во сотрудников у менеджера
pd.read_sql('SELECT ps."Manager Name" AS manager, COUNT(ps."Employee Name") AS ppl_count\
             FROM production_staff AS ps\
             GROUP BY manager ORDER BY ppl_count DESC', connect)

manager  ppl_count
0   Michael Albert         23
1     Elijiah Gray         23
2   Kissy Sullivan         22
3    Kelley Spirea         22
4    David Stanley         21
5    Ketsia Liebig         21
6   Brannon Miller         21
7         Amy Dunn         21
8   Webster Butler         21
9   Elisa Bramante         14
10            None          0

In [40]:
# Зависимость кол-ва увольнений от возраста
pd.read_sql('SELECT ps."Reason for Term" AS reason, MIN(hr."age") AS min_age, ROUND(AVG(hr."age")) AS avg_age,\
             MAX(hr."age") AS max_age\
             FROM production_staff AS ps\
             JOIN hr_dataset AS hr\
             ON ps."Employee Name" = hr."Employee Name"\
             WHERE ps."Employment Status" = \'Voluntarily Terminated\'\
             GROUP BY reason ORDER BY avg_age DESC',connect)

reason  min_age  avg_age  max_age
0                           retiring       35     52.0       65
1                         attendance       44     44.0       44
2                           military       26     44.0       63
3                   Another position       27     41.0       65
4                         more money       30     41.0       67
5                      career change       30     41.0       53
6                            unhappy       28     41.0       66
7             relocation out of area       34     40.0       48
8                              hours       29     39.0       48
9                   return to school       28     37.0       53
10                    medical issues       31     31.0       31
11  maternity leave - did not return       30     31.0       31

In [41]:
# Зависимость семейного положения и количества увольнений
pd.read_sql('SELECT hr."maritaldesc" AS marital_status, COUNT(ps."Reason for Term") AS count\
             FROM hr_dataset AS hr\
             JOIN production_staff AS ps\
             ON ps."Employee Name" = hr."Employee Name"\
             WHERE ps."Employment Status" = \'Voluntarily Terminated\'\
             GROUP BY marital_status ORDER BY count DESC',connect)

marital_status  count
0        Married     34
1         Single     23
2       Divorced     14
3        Widowed      3
4      Separated      1

### Вывод:
   По полученым результатам видно, что чаще всего сотрудники увольняются из отдела *'Production'* - 75 человек. При этом наблюдается зависимость между менеджером сотрудника и количеством увольнений.<br>  
У менеджеров **'Amy Dunn'** и **'Webster Butler'** по 13 уволившихся  сотрудника (за каждым было закреплено по 21 человеку).
Отдельно стоит обратить внимание на менеджера **'Ketsia Liebig'**. За ней также был закреплен 21 человек, но количество уволившихся - 4 человека. <br>

Не мало важен и источник найма:
   - 'Diversity Job Fair' - 29 человек нанято, 16 уволились (58% от нанятых)
   - 'Search Engine - Google Bing Yahoo' - 25 человек нанято, 14 уволились (56% от нанятых)
   - 'Monster.com - 24 человека нанято, 10 уволились (42% от нанятых)<br>
   
И наоборот:
   - 'Pay Per Click - Google' - 21 человек нанято, 3 уволились (15% от нанятых)
   - 'Professional Society' - 20 человек нанято, 2 уволились (10% от нанятых)<br>
 
Возраст и семейное положение никак не влияют на количество увольнений.      

### Предмет исследования №3:
### Days Employed - Количество отработаных дней 

In [46]:
# Минимальная, средняя и максимальная продолжительность работы и возраст сотрудников в отделах
pd.read_sql('SELECT hr."department" AS dept, COUNT(hr."Employee Name") AS ppl_count,\
             MIN(hr."Days Employed") AS min_days, MAX(hr."Days Employed") AS max_days,\
             ROUND(AVG(hr."Days Employed")) AS avg_days, MIN(hr."age") AS min_age,\
             ROUND(AVG(hr."age")) AS avg_age, MAX(hr."age") AS max_age\
             FROM hr_dataset AS hr\
             GROUP BY dept ORDER BY avg_days DESC', connect)

dept  ppl_count  min_days  max_days  avg_days  min_age  \
0      Executive Office          1      1973      1973    1973.0       63   
1                 Sales         31       425      4339    1622.0       27   
2         Admin Offices         10        58      3317    1421.0       30   
3  Software Engineering         10       421      2295    1418.0       30   
4     Production               208         2      3674    1308.0       25   
5                 IT/IS         50         6      2787     982.0       28   

   avg_age  max_age  
0     63.0       63  
1     39.0       63  
2     32.0       39  
3     36.0       51  
4     39.0       67  
5     38.0       66

In [47]:
# Зависимость семейного положения и продолжительности работы
pd.read_sql('SELECT hr."maritaldesc" AS marital_status, MIN(hr."Days Employed") AS min_days,\
             MAX(hr."Days Employed") AS max_days,ROUND(AVG(hr."Days Employed")) AS avg_days\
             FROM hr_dataset AS hr\
             GROUP BY marital_status ORDER BY avg_days DESC', connect)

marital_status  min_days  max_days  avg_days
0      Separated       905      4339    1703.0
1         Single         8      3674    1373.0
2        Married         2      3611    1246.0
3       Divorced         8      2686    1103.0
4        Widowed         2      1602     863.0

In [48]:
# Зависимость кто менеджер и продолжительность работы сотрудника
pd.read_sql('SELECT hr."Manager Name" AS manager, MIN(hr."Days Employed") AS min_days,\
             MAX(hr."Days Employed") AS max_days, ROUND(AVG(hr."Days Employed")) AS avg_days\
             FROM hr_dataset AS hr\
             GROUP BY manager ORDER BY avg_days DESC', connect)

manager  min_days  max_days  avg_days
0   Board of Directors      1973      3247    2610.0
1         Eric Dougall      1908      2766    2384.0
2        Lynn Daneault       508      4339    1777.0
3      Jennifer Zamora       446      2417    1694.0
4           Janet King        98      3247    1620.0
5           John Smith       425      2617    1599.0
6       Brannon Miller       399      3065    1592.0
7        Ketsia Liebig       124      2771    1490.0
8        Kelley Spirea       164      2645    1387.0
9         Elijiah Gray         8      3611    1374.0
10       David Stanley         2      3674    1370.0
11     Alex Sweetwater       421      2148    1320.0
12  Brandon R. LeBlanc        58      3317    1233.0
13      Debra Houlihan       899      1301    1163.0
14      Webster Butler        57      2190    1125.0
15      Michael Albert        69      2428    1098.0
16            Amy Dunn         2      2148    1078.0
17      Kissy Sullivan        19      2512    1047.0
18        Peter Monroe       127      1153     943.0
19          Simon Roup         6      1112     679.0
20    Brian Champaigne       220       420     308.0

In [49]:
# Распределение менеджеров по отделам
pd.read_sql('SELECT hr."Manager Name" AS manager, hr."department" AS dept\
             FROM hr_dataset AS hr\
             GROUP BY manager, dept ORDER BY manager', connect)

manager                  dept
0      Alex Sweetwater  Software Engineering
1             Amy Dunn     Production       
2   Board of Directors         Admin Offices
3   Board of Directors      Executive Office
4   Brandon R. LeBlanc         Admin Offices
5       Brannon Miller     Production       
6     Brian Champaigne                 IT/IS
7        David Stanley     Production       
8       Debra Houlihan                 Sales
9         Elijiah Gray     Production       
10        Eric Dougall                 IT/IS
11          Janet King         Admin Offices
12          Janet King     Production       
13          Janet King                 IT/IS
14          Janet King                 Sales
15     Jennifer Zamora  Software Engineering
16     Jennifer Zamora                 IT/IS
17          John Smith                 Sales
18       Kelley Spirea     Production       
19       Ketsia Liebig     Production       
20      Kissy Sullivan     Production       
21       Lynn Daneault                 Sales
22      Michael Albert     Production       
23        Peter Monroe                 IT/IS
24          Simon Roup                 IT/IS
25      Webster Butler     Production

In [50]:
# Зависимость продолжительности работы от пола сотрудника
pd.read_sql('SELECT hr."sex" AS sex, MIN(hr."Days Employed") AS min_days,\
             MAX(hr."Days Employed") AS max_days, ROUND(AVG(hr."Days Employed")) AS avg_days\
             FROM hr_dataset as hr\
             GROUP BY sex',connect)

sex  min_days  max_days  avg_days
0  Female         2      3611    1313.0
1    Male         2      4339    1273.0

In [53]:
# Зависимость продолжительности работы от расы сотрудника
pd.read_sql('SELECT hr."racedesc" AS race, MIN(hr."Days Employed") AS min_days,\
             MAX(hr."Days Employed") AS max_days, ROUND(AVG(hr."Days Employed")) AS avg_days\
             FROM hr_dataset as hr\
             GROUP BY race ORDER BY avg_days DESC',connect)

race  min_days  max_days  avg_days
0  American Indian or Alaska Native      1196      2336    1845.0
1                 Two or more races        45      3065    1628.0
2                          Hispanic         2      2148    1547.0
3         Black or African American         6      3674    1436.0
4                             White         2      4339    1230.0
5                             Asian        19      2583    1164.0

### Вывод:

При выявлении зависимостей я не учитывал:<br>
   - отдел Executive Officе<br>
   - менеджеров:<br>
       - Board of Directors<br>
       - Janet King<br>*Janet King занимает должность President & CEO в отделе Executive Officе, в котором она единственный сотрудник. Board of Directors (совет директоров) является менеджером Janet King.*<br>

Из полученных данных видно, что средняя продолжительность работы сотрудника зависит от отдела в котором он работает. В отделе 'Sales' - средняя продолжительность работы составляет 1622 дня(максимальный показатель). А в отделе 'IT/IS' - средняя продолжительность - 982 дня(минимальный показатель).<br>

Также наблюдается зависимость продолжительности работы сотрудника от того, кто у него менеджер.
Под руковдством **Eric Dougall** в среднем люди работают - 2384 дня. В то же время у **Brian Champaigne** средняя продолжительность работы сотрудников - 308 дня. Интересным является тот факт, что оба менеджера работают в отделе 'IT/IS'.<br>

Семейное положение является одним из факторов влияющих на продолжительность работы. Те у кого в графе семейное положение стоит 'Separated' имеют самые выскоие показатели по среднему количеству отработанных дней - 1703.<br>

Раса также оказывает влияние.Представители 'American Indian or Alaska Native' в среднем работают 1845 дней. А у 'Asian' самый низкий показатель - 1164 дня.
Ни пол ни ворзраст сотрудника на продолжительность работы не влияют.